### ISSUE TITLE
unhandled error

### ISSUE THREAD

**Issue by derimmer** [vasl-developers/vasl#811]
from vassal bug report 13624 and 13627:

Attempting to load a board.

VASL v6.6.1 3.4.8

java.lang.NullPointerException: Cannot invoke "VASL.build.module.map.boardArchive.DataFile.getVersion()" because "this.dataFile" is null
 at VASL.build.module.map.boardArchive.BoardArchive.getVersion(BoardArchive.java:786)
 at VASL.build.module.map.boardPicker.ASLBoard.initializeFromArchive(ASLBoard.java:139)
 at VASL.build.module.map.ASLBoardPicker.updateBoard(ASLBoardPicker.java:508)
 at VASL.build.module.map.ASLBoardPicker.getBoard(ASLBoardPicker.java:601)
---------------

Trying to open a VSAV file

VASL v6.6.1 3.4.8

java.lang.NullPointerException: Cannot invoke "VASL.build.module.map.boardArchive.BoardArchive.getBoardName()" because "this.VASLBoardArchive" is null
 at VASL.build.module.map.boardPicker.ASLBoard.initializeFromArchive(ASLBoard.java:137)
 at VASL.build.module.map.ASLBoardPicker.updateBoard(ASLBoardPicker.java:508)
 at VASL.build.module.map.ASLBoardPicker.decode(ASLBoardPicker.java:139)
 at VASSAL.launch.BasicModule.decodeSubCommand(BasicModule.java:238)

**Comment by derimmer** [comment]
Here is another error with the same root problem. VASSAL bug report 13857


java.lang.NullPointerException: Cannot invoke "VASL.build.module.map.boardArchive.BoardArchive.isLegacyBoard()" because "this.VASLBoardArchive" is null  at VASL.build.module.map.boardPicker.ASLBoard.isLegacyBoard(ASLBoard.java:181)  at VASL.build.module.map.boardPicker.VASLBoard.getName(VASLBoard.java:139)  at VASL.build.module.ASLMap.setBoards(ASLMap.java:182)  at VASSAL.build.module.map.BoardPicker.setup(BoardPicker.java:483)
--

All of these errors seem to be build around VASLBoardArchive being null; a problem I cannot replicate. 

@uckelman you had helped me before (last fall I think) with some messages about error handling which I cannot find at all. Can I prevail on your patience for some more help. 

It seems to me that the problem lies here in ASLBoard.java:

public void initializeFromArchive(File archiveFile) {

        try {
            VASLBoardArchive = new BoardArchive(archiveFile.getName(), archiveFile.getParent(), ASLMap.getSharedBoardMetadata());
        } catch (IOException e) {
            ErrorDialog.dataWarning(new BadDataReport("Unable to open board file", archiveFile.getName(), e));

        }
        try {
            boardFile = archiveFile;
            setCommonName(VASLBoardArchive.getBoardName());
            imageFile = VASLBoardArchive.getBoardImageFileName();
            version = VASLBoardArchive.getVersion();

            ((Translatable) getGrid()).setAttribute(HexGrid.X0, (int) VASLBoardArchive.getA1CenterX());
            ((Translatable) getGrid()).setAttribute(HexGrid.Y0, (int) VASLBoardArchive.getA1CenterY());
            ((Translatable) getGrid()).setAttribute(HexGrid.DX, VASLBoardArchive.getHexWidth());
            ((Translatable) getGrid()).setAttribute(HexGrid.DY, VASLBoardArchive.getHexHeight());
            ((Translatable) getGrid()).setAttribute(HexGrid.SNAP_SCALE, VASLBoardArchive.getSnapScale());
            ((Translatable) getGrid()).setAttribute(BoardArchive.ALT_HEX_GRID_KEY, Boolean.toString(VASLBoardArchive.isAltHexGrain()));
        } catch (Exception e) {
            // Fail silently
            return ;
        }
    }

The IOException here should get called if VASLBoardArchive is null but it simply reports out an error and lets the code continue which then triggers an error in routines that call initializeFromArchive(). Should we be handling the error differently here or in the calling code?

**Comment by zgrose** [comment]
One problem is you're creating a BoardArchive here `VASLBoardArchive = new BoardArchive(archiveFile.getName(), archiveFile.getParent()...` and then using the VASLBoardArchive later e.g. `setCommonName(VASLBoardArchive.getBoardName());` even if the constructor failed in your first try/catch block.

At a bare minimum I would refactor this function as

````
public void initializeFromArchive(File archiveFile) {

    try {
        VASLBoardArchive = new BoardArchive(archiveFile.getName(), archiveFile.getParent(), ASLMap.getSharedBoardMetadata());
    } catch (IOException e) {
        ErrorDialog.dataWarning(new BadDataReport("Unable to open board file", archiveFile.getName(), e));
        return;
    }

    ....
````

If the VASLBoardArchive isn't initialized, there really isn't much reason to do anything else inside of `initializeFromArchive`

Also, why not do another `ErrorDialog.dataWarning(....)` inside the final catch block at the bottom?

**Comment by derimmer** [comment]
Thanks @zgrose. I have implemented both of your suggestions. Am also going to look at where initializeFromArchive is called and see if I can trap the error there and prevent a VASSAL bug.

**Comment by derimmer** [comment]
@uckelman I found the advice that you had given my previously last October (see issue #773 and PR #775). I will proceed as per your suggestions. A lot of our VASSAL bug reports stem from ASLBoardPicker, mainly due to incorrect user input which we are not managing.

**Comment by uckelman** [comment]
What @zgrose says above is right. If the rest of that function requires a nonnull `VASLBoardArchive`, then you have to bail out before reaching that point if you get an `IOException` when trying to create one.

**Comment by derimmer** [comment]
> What @zgrose says above is right. If the rest of that function requires a nonnull `VASLBoardArchive`, then you have to bail out before reaching that point if you get an `IOException` when trying to create one.

Thanks. Please see my comments under issue #773.

**Comment by derimmer** [comment]
OK. We have tried to fix this with PR #868. See also #773
